src/Chapter3.hs

```diff
@@ -480,6 +531,55 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data NumPeople = One | Two | Three | Four
+              deriving (Show, Enum)
+
+data Castle = Castle String
+            deriving (Show)
+
+data Wall = Wall Castle
+        deriving (Show)
+
+data Institution = Church | Library
+                deriving (Show)
+
+data House = House NumPeople
+          deriving (Show)
+
+data City =   CityWithCastle Castle Institution [House]
+            | WalledCity Wall Institution [House]
+            | City Institution [House]
+            deriving (Show)
```

Comment:
That is an interesting representation of the city as a sum type 🏙️ 

However, you can see a lot of repetition in all constructors.
I feel like it would be more convenient to work `City` if it was a record type instead of a sum type. 

First of all, if you change the castele to have walls inside, it would then be possible to only have the castle represented with three states in one record type rather than three constructors with almost the same types. Let's think about how to do that. The description says:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type indeed!

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. These zrfe just the suggestions and simplifications, or just exploration of different approaches 🙂 

So, instead of having three top-level constructors and duplicating all fields on both of them, you can define `City` as a record with the new castle:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- Church or library but not both
    , cityInstitution :: Institution
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouses :: [House]
    }
```